CYS - Core: install font when user clicks opt-in

gigitux opened the pull request:
> [!WARNING]  
> This PR is blocked by #45586

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #45476 .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure you have the WooCommerce Beta Tester plugin installed and activated
2. Under Tools > WCA Test Helper > Features, make sure `customize-store` is enabled
3. Under Tools > WCA Test Helper > Tools, click on `Reset Customize Your Store`-
5. Visit `wp-admin/admin.php?page=wc-settings&tab=advanced&section=woocommerce_com`
6. Ensure that the `Allow usage of WooCommerce to be tracked` is disabled.
7. Visit `wp-admin/admin.php?page=wc-admin&path=%2Fcustomize-store`.
8. Click on start `Start designing`.
9. On the sidebar, click on `Change homepage`.
10. Select a new homepage.
11. Save the change.
12. On the sidebar, click on `Choose Fonts`.
13. Check `Opt in to usage tracking to get access to more fonts`.
14. Click on `usage tracking`.
15. Click on Opt-in.
16. Ensure that the pre-load screen appears
17. Ensure that the Assembler is loaded
18. Click on the "Choose Fonts"
19. Ensure that multiple fonts are available.
20. Ensure that the homepage hasn't been restored.
21. Install https://wordpress.org/plugins/wp-data-access/
22. Visit wp-admin
23. On the sidebar WP Data Access > Query Builder
24. Copy paste: 
```
SELECT * FROM `wp_posts` WHERE `post_type` LIKE 'wp_font_family' ```

25. Click Execute
26. Ensure that a table with some results is visible
27. Copy paste:
```
SELECT * FROM `wp_posts` WHERE `post_type` LIKE 'wp_font_face' 
```
28. Ensure that a table with some results is visible
29. Remove them via:
```
DELETE FROM `wp_posts` WHERE `post_type` = 'wp_font_face';
DELETE FROM `wp_posts` WHERE `post_type` = 'wp_font_family';
```
30. Ensure that the `woocommerce_allow_tracking` option (`wp_option` table) is set to `yes`.
31. Update the option to `no`.
32. Refresh the assembler page.
33. Repeat 9-20.


<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
CYS - Core: install font when user clicks opt-in

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->


</details>

gigitux committed:
CYS - Core: install font when user clicks opt-in

gigitux committed:
remove debugger

gigitux committed:
CYS - Core: fix wp-admin page visible when click on start designing

gigitux committed:
Merge branch '45542-cys-core-wp-admin-screen-visible-for-a-moment-after-clicking-start-designing-1' of github.com:woocommerce/woocommerce into 45476-cys-on-core-make-sure-the-pre-loader-is-triggered-right-after-the-user-clicks-on-the-opt-in-modal

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

gigitux committed:
use sendEvent instead of redirect

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

gigitux committed:
not show wp-admin after the click on the dialog button

gigitux committed:
Merge branch '45542-cys-core-wp-admin-screen-visible-for-a-moment-after-clicking-start-designing-1' of github.com:woocommerce/woocommerce into 45476-cys-on-core-make-sure-the-pre-loader-is-triggered-right-after-the-user-clicks-on-the-opt-in-modal

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into 45476-cys-on-core-make-sure-the-pre-loader-is-triggered-right-after-the-user-clicks-on-the-opt-in-modal

gigitux committed:
use sendEventToIntroMachine

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into 45476-cys-on-core-make-sure-the-pre-loader-is-triggered-right-after-the-user-clicks-on-the-opt-in-modal

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into 45476-cys-on-core-make-sure-the-pre-loader-is-triggered-right-after-the-user-clicks-on-the-opt-in-modal

gigitux committed:
fix font installation

gigitux commented:
@albarin, @nefeline and @verofasulo. This PR introduces this UX:
https://github.com/woocommerce/woocommerce/assets/4463174/21d85b46-8d54-4e04-8106-e6d08c54fb89


Please note that after installation, clicking on "Choose Fonts" will not select any font pair. This is because the theme continues to use the previously chosen font by default:
![image](https://github.com/woocommerce/woocommerce/assets/4463174/d29842ff-eb97-4112-be50-0b742c3f679d)

Should we consider updating the theme to utilize the newly installed font (Inter -Inter) during the installation process?

gigitux committed:
fix eslint error

danielwrobert commented on plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/sidebar-navigation-screen-typography.tsx:
```diff
@@ -201,18 +202,16 @@ export const SidebarNavigationScreenTypography = () => {
 										</Button>
 										<Button
 											onClick={ async () => {
```
Is the `async` still needed here or can this be a regular arrow fn now that we removed the `await` portion below?

gigitux replied:
Nice catch! Fixed!

danielwrobert commented:
**Nit:** I'm not sure if it's appropriate to address in this PR but it seems like the `Opt in to` portion of the `Opt in to usage tracking to get access to more fonts.` should also be included in the link so it's a more clear CTA.

![CleanShot 2024-03-15 at 17 10 00](https://github.com/woocommerce/woocommerce/assets/481776/223752e5-daf2-4f31-962e-f50417b7a62d)

Having `usage tracking` be the linked text alone makes me think I will be directed to information about the usage tracking.

danielwrobert reviewed (commented):
> Ensure that the pre-load screen appears

I've tried to run the testing steps 4 times and each time it freezes at the  **Turning on the lights** step.

I can move forward in the process by refreshing the page but if I don't refresh, it will just hang.

https://github.com/woocommerce/woocommerce/assets/481776/643d9399-7f5f-41e6-a6d4-ded43959420e

nefeline commented:
> Nit: I'm not sure if it's appropriate to address in this PR but it seems like the Opt in to portion of the Opt in to usage tracking to get access to more fonts. should also be included in the link so it's a more clear CTA. Having usage tracking be the linked text alone makes me think I will be directed to information about the usage tracking.

@danielwrobert could you please direct any questions related to design and/or copy to @verofasulo and @lauroraa ? Thanks!

nefeline commented on plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/sidebar-navigation-screen-typography.tsx:
```diff
@@ -201,18 +202,16 @@ export const SidebarNavigationScreenTypography = () => {
 											onClick={ async () => {
-												await dispatch(
-													OPTIONS_STORE_NAME
-												).updateOptions( {
```
Prior to this change, we were indirectly triggering synchronous font installation by triggering the state whenever accessing the link to the assembler. Now, you created a dedicated event for this, which is great! With that said, can you please re-add this async font installation here? It's important to have **both the sync and async fonts installation as a safety net** in case one of them fails to execute for some reason.

gigitux replied:
> Prior to this change, we were indirectly triggering synchronous font installation by triggering the state whenever accessing the link to the assembler. Now, you created a dedicated event for this, which is great! With that said, can you please re-add this async font installation here? It's important to have both the sync and async fonts installation as a safety net in case one of them fails to execute for some reason.

I just moved the logic [here](https://github.com/woocommerce/woocommerce/pull/45580/files#diff-10fabebab9142a466545a1dc81f748b9cbf82bff2fc280e5aee18e1629a0c900R238-R248)

nefeline reviewed (changes requested):
Thanks for your work here, @gigitux! I've requested a change in the code: let's keep both async and sync fonts installation in place in case one of them happen to fail 👍 

Regarding:

> and each time it freezes at the Turning on the lights step.

I can reproduce this as well: unfortunately, the console has no particular errors that could help with debugging.

albarin commented:
> I've tried to run the testing steps 4 times and each time it freezes at the **Turning on the lights** step.

I can reproduce this as well.

gigitux committed:
fix install font phase when the iframe is not loaded

gigitux committed:
remove not necessary async

gigitux commented:
Sorry for the confusion, I introduced the issue with [00eb4a3](https://github.com/woocommerce/woocommerce/pull/45580/commits/00eb4a34efe02a10bd4601301217687e98b096f1).

I fixed it! Also, I noticed that the font installation wasn't triggered when the assembler wasn't loaded in the iframe. I fixed that issue as well! 🤞

albarin reviewed (approved):
Great effort 🙇 
Tests as expected now!

nefeline reviewed (approved):
Thanks for the amazing work! Confirmed things are working as expected now 🚀
